AK/BigIntBase.h

```diff
@@ -0,0 +1,566 @@
+            output = __builtin_addcll(word1, word2, carry, reinterpret_cast<unsigned long long*>(&ncarry));
+        else
+            VERIFY_NOT_REACHED();
+        carry = ncarry;
+        return;
+#    endif
+#endif
+    }
+    // Note: This is usually too confusing for both GCC and Clang.
+    bool ncarry = __builtin_add_overflow(word1, word2, &output);
+    if (carry) {
+        ++output;
+        if (output == 0)
+            ncarry = true;
+    }
+    carry = ncarry;
+}
+
+ALWAYS_INLINE constexpr void sub_words(NativeWord word1, NativeWord word2, NativeWord& output, bool& carry)
+{
+    if (!__builtin_is_constant_evaluated()) {
+#if ARCH(X86_64)
+#    if defined(AK_COMPILER_CLANG)
+        // This is named __builtin_ia32_sbb_u64 in G++. *facepalm*
+        carry = __builtin_ia32_subborrow_u64(carry, word1, word2, reinterpret_cast<unsigned long long*>(&output));
+#    else
+        carry = __builtin_ia32_sbb_u64(carry, word1, word2, reinterpret_cast<unsigned long long*>(&output));
+#    endif
+        return;
+#else
+#    if defined(AK_COMPILER_CLANG)
+        NativeWord ncarry;
+        if constexpr (SameAs<NativeWord, unsigned int>)
+            output = __builtin_subc(word1, word2, carry, reinterpret_cast<unsigned int*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long>)
+            output = __builtin_subcl(word1, word2, carry, reinterpret_cast<unsigned long*>(&ncarry));
+        else if constexpr (SameAs<NativeWord, unsigned long long>)
+            output = __builtin_subcll(word1, word2, carry, reinterpret_cast<unsigned long long*>(&ncarry));
+        else
+            VERIFY_NOT_REACHED();
+        carry = ncarry;
+        return;
+#    endif
+#endif
+    }
+    bool ncarry = __builtin_sub_overflow(word1, word2, &output);
+    if (carry) {
+        if (output == 0)
+            ncarry = true;
+        --output;
+    }
+    carry = ncarry;
+}
+
+// Calculate ((divident1 << word_size) + divident0) / divisor. Quotient should be guaranteed to fit into NativeWord.
+ALWAYS_INLINE constexpr NativeWord div_mod_words(NativeWord divident0, NativeWord divident1, NativeWord divisor, NativeWord& remainder)
+{
+#if ARCH(x86_64)
+    if (!__builtin_is_constant_evaluated()) {
+        NativeWord remainder_value, quotient;
+        asm("div %3"
+            : "=a"(remainder_value), "=d"(quotient)
+            : "a"(divident0), "d"(divident1), "rm"(divisor));
+        remainder = remainder_value;
+        return quotient;
+    }
+#endif
+    auto divident = (static_cast<DoubleWord>(divident1) << word_size) + divident0;
+    remainder = static_cast<NativeWord>(divident % divisor);
+    return static_cast<NativeWord>(divident / divisor);
+}
+
+// ===== Operations on integer storages =====
+constexpr void storage_copy(auto const& operand, auto&& result, size_t offset = 0)
+{
+    auto fill = extend_sign(operand.is_negative());
+    size_t size1 = operand.size(), size = result.size();
```

Comment:
itʼd be good to have longer names, even `operand_size` and `result_size` (as far as i know, comma operator is not really popular here…)

Reply:
I can do this but trailing digits actually trivially encode the information you ask me to add: 1 means belonging to `operand1` (or the only operand), 2 means belonging to `operand2`, nothing &mdash; belonging to result. I might just add this note somewhere to the file.

Reply:
hmm… up to your discretion then! (here we would need to use `operand1` as a name to comply, but i donʼt think its worth it; yet longer names wouldnʼt compile to a longer code… 🙃)